src/comm.cc

```diff
@@ -385,6 +385,7 @@ comm_openex(int sock_type,
         AI->ai_protocol = proto;
         debugs(50, 3, "Attempt fallback open socket for: " << addr );
         new_socket = socket(AI->ai_family, AI->ai_socktype, AI->ai_protocol);
+        xerrno = errno;
```

Comment:
I am not 100% sure that this code is meant to preserve the _last_ error when trying a second socket() system call after the first failure. If you are sure that this is wrong, please let me know what the replacement C++ comment should say:

```suggestion
        // keep xerrno unchanged because ...XXX...
```

Reply:
IIRC this is preserving the *first* errno, so the primary socket failure can be fixed, not just the fallback type.

That said we have had several cycles of fixing errno usage, so we should now check whether the callers of this function actually need it preserved. Perhapse this kludge can all go away.

Reply:
> IIRC this is preserving the _first_ errno, so the primary socket failure can be fixed, not just the fallback type.

I changed the code to preserve the first errno in commit 5dfa03d.

While working on that change, I discovered another bug in this code. There is no trivial or in-scope fix, so I added an XXX (while preserving the current broken official code behavior).
 
> That said we have had several cycles of fixing errno usage, so we should now check whether the callers of this function actually need it preserved. Perhapse this kludge can all go away.

Unfortunately, there are still many (indirect) comm_openex() callers that use `errno`.

Comm should stop setting `errno` for callers (and stop returning invalid descriptor numbers on errors), but that requires a lot of changes outside this PR scope, of course. It can probably be done incrementally, but the first step needs to set a proper framework up. This small PR cannot do that.